dra-evolution/pkg/api/capacity_types.go

```diff
@@ -3,168 +3,110 @@ package api
-	// that can be allocated in a manner similar to standard Kubernetes
-	// resources.
-	//
-	// +optional
-	ClaimResourcesProvided []ResourceCapacity `json:"claimResourcesProvided,omitempty"`
+	// TODO for 1.31: define how to support partitionable devices
 }
 
-type ResourceCapacity struct {
-	// Name is the resource name/type.
-	// +required
-	Name string `json:"name"`
-
-	// Capacity is the total capacity of the named resource.
-	// +required
-	Capacity resource.Quantity `json:"capacity"`
-
-	// BlockSize is the increments in which capacity is consumed. For
-	// example, if you can only allocate memory in 4k pages, then the
-	// block size should be "4Ki". Default is 1.
+// DeviceAttribute is a combination of an attribute name and its value.
+type DeviceAttribute struct {
+	// Name is a unique identifier across all possible attributes of devices.
+	// It must be a DNS subdomain, with one additional restriction: the
+	// first part must not contain a hyphen and not start with a digit.
 	//
-	// If the resource is consumable in a fractional way, then the
-	// default of 1 should not be used; instead this should be a fractional
-	// amount corresponding the increment size. We may also need a minimum
-	// value, if the minimum is larger than the block size (as is the case
-	// for standard Kubernetes CPU resources).
+	// If this is a DNS label (no dot), then the driver name gets added
+	// when looking up attributes. This avoids name collisions with attributes
+	// used by other drivers.
 	//
-	// +optional
-	BlockSize *resource.Quantity `json:"blockSize,omitempty"`
-}
+	// If this is a full DNS subdomain, then the meaning of the attribute is driver-independent.
+	// For example, Kubernetes will use `*.k8s.io` names when defining attributes that
+	// drivers from different vendors are supposed to use.
+	Name string `json:"name" protobuf:"bytes,1,name=name"`
 
-// Attribute capture the name, value, and type of an device attribute.
-type Attribute struct {
-	Name string `json:"name"`
-
-	// One of the following:
-	StringValue   *string            `json:"stringValue,omitempty"`
-	IntValue      *int               `json:"intValue,omitempty"`
-	QuantityValue *resource.Quantity `json:"quantityValue,omitempty"`
-	SemVerValue   *SemVer            `json:"semVerValue,omitempty"`
+	DeviceAttributeValue `json:",inline" protobuf:"bytes,2,opt,name=attributeValue"`
 }
 
-func (a Attribute) Equal(b Attribute) bool {
-	if a.Name != b.Name {
-		return false
-	}
-
-	return a.EqualValue(b)
+// The Go field names below have a Value suffix to avoid a conflict between the
+// field "String" and the corresponding method. That method is required.
+// The Kubernetes API is defined without that suffix to keep it more natural.
+
+// DeviceAttributeValue must have one and only one field set.
+type DeviceAttributeValue struct {
+	// QuantityValue is a quantity.
+	QuantityValue *resource.Quantity `json:"quantity,omitempty" protobuf:"bytes,6,opt,name=quantity"`
```

Comment:
These could be pulled up into the `DeviceAttribute` type - any reason not to?

Reply:
I've mentioned this a few times in other threads, but maybe it keeps getting lost...

The nice thing about keeping embedded structs is that you can add the following to such structs to _enforce_ their one-of semantics.
```
// +kubebuilder:validation:MaxProperties=1
```
For example ...
```
// DeviceAttribute is a combination of an attribute name and its value.
type DeviceAttribute struct {
	Name string `json:"name"
	DeviceAttributeValue `json:",inline"
}

// DeviceAttributeValue must have one and only one field set.
// +kubebuilder:validation:MaxProperties=1
type DeviceAttributeValue struct {
	QuantityValue *resource.Quantity `json:"quantity,omitempty"
	BoolValue *bool `json:"bool,omitempty"
	StringValue *string `json:"string,omitempty"
	VersionValue *string `json:"version,omitempty"
}
```

The following looks nicer in go, but you lose the ability to force users to only set one of the value types.
```
// DeviceAttribute is a combination of an attribute name and its value.
type DeviceAttribute struct {
	Name string `json:"name"

	QuantityValue *resource.Quantity `json:"quantity,omitempty"
	BoolValue *bool `json:"bool,omitempty"
	StringValue *string `json:"string,omitempty"
	VersionValue *string `json:"version,omitempty"
}
```

Reply:
See https://github.com/kubernetes-sigs/wg-device-management/pull/14#discussion_r1623902425.

Reply:
I haven't added the `+oneof` but made the other change.